### ISSUE TITLE
Video scrolling laggy

### ISSUE THREAD

**Issue by MartinDelille** [Phonations/joker#398]
Since new video thread playback

**Comment by tlecomte** [comment]
Hi Martin! I just tried Joker 2.1.3 (did not try the new versions from a long time...) and indeed there is an issue with the scrolling that was not there in the PR at first...

I believe the problem lies in this commit:
https://github.com/Phonations/Joker/pull/392/commits/4f7809d5ccf7b02d2e7a285c4761341a5e1b19fa

I did not notice that when you asked for comments... The goal of the recursive design was precisely to provide fast scrolling. When the user scrolls, there is a massive amount of frames being requested, as many as the positions taken by the cursor in the time bar. Trying to decode them all produces a significant lag. The replacement with a processing while loop is a 99% equivalent of the recursive design, except **the 5 ms sleep**... This sleep combined with the frame cancellation criteria, which depends in the frame pool size, most likely is the cause of the lags.

- PhVideoPool.cpp, line 174: during the scroll, each time cleanup() is called, we emit a cancelFrameRequest signal for all frames that have been requested, are not yet decoded, and are no longer needed given the current time. So the scrolling will generate a large amount of CancelFrameRequest signals. However, at the end of the scrolling, the criteria for cancelling a frame no longer applies:

```
    if ((bufferFrame < frame - halfPoolSize) || (bufferFrame >= frame + halfPoolSize)) {
     // Given the current position,
     // this frame is not likely to be shown on screen anytime soon.
```

At the end of the scroll we can have up to halfPoolSize requested frames that will not be cancelled.

So the lag can be up to (5ms)*halfPoolSize...

However I guess you cannot remove that sleep time, or you end up with a busy loop...

One option is to go back to the recursive design which has no sleep time issue... However, you said that there was sometimes a stack overflow. Maybe` _recursive` could be made an integer, and if it gets larger that a certain level then we stop the recursivity:

```
   // call processEvents to walk through the queued signals
   // This makes sure we process the cancelling signals.
   // We use the _recursive flag to indicate that the child slots
   // should not actually decode the frame.
   _recursive += 1;
   if (_recursive > 100) {
       QCoreApplication::processEvents();
   }
   _recursive -= 1;
```

How does that sound?

P.S.: the whole reason for this gymnastic is that we use signal/slots to manage the list of requested frames, instead of a shared variable protected by a lock. But I still think that signal/slots are less dangerous that locks...

**Comment by tlecomte** [comment]
Another possibility: just do not do the 5ms sleep if _requestedFrames is not empty.

**Comment by MartinDelille** [comment]
Hi @tlecomte ! Good to see you there ;-)
I didn't liked the recursive design at the begginning but adding the protection might be fine.

**Comment by tlecomte** [comment]
I thought about this a bit more and the solution is actually simpler...

The real root of the stack overflow problem did not come from the `QCoreApplication::processEvents()` call, since this part is not recursive! It actually come from the following part that was at the end of `decodeFrame()`:

	// call the same function again to proceed with the other requested frames
	if (!_requestedFrames.empty()) {
		decodeFrame(NULL);
	}

So the proper solution is just to replace this recursive call to `decodeFrame `by a while loop on `_requestedFrames`. And `QCoreApplication::processEvents()` could also be called at each step of this loop.

On the other hand, I also realize that the `_recursive` variable is useless, it can be replaced by `!_requestedFrames.empty()`.

I was going to make a PR for these changes, but unfortunately I upgraded to Qt 5.5 on my machine, and Joker does not build anymore. There are linking errors on OpenGL functions...

**Comment by MartinDelille** [comment]
Hi @tlecomte,

I'll be happy to see your PR if you manage to build Joker again. I add a Qt5.5 build job to appveyor to see if it reproduces your problem: #401

Sometime with Qt, it might be safe to totally remove and reinstall the environment.

**Comment by MartinDelille** [comment]
@xela13 Il y a un truc qui devrait permettre de réduire les saccades: 

1. Ouvrir la base de registre (_Démarrer > Exécuter... > regedit.exe_)
2. Naviguez vers la clé \HKEY_CURRENT_USER\Software\Phonations\Joker
3. Ajoutez deux valeurs de type entier:
  - videoReadHead : 5
  - videoPoolSize : 10

**Comment by xela13** [comment]
@MartinDelille 
j'ai essayé mais sans succès. je t'ai tout expliqué sur le messenger facebook.

**Comment by MartinDelille** [comment]
@xela13 J'ai mis à jour mon message pour corriger le problème

**Comment by tlecomte** [comment]
@MartinDelille I came back to this issue after some months, and with some profiling, I was able to remove a massive performance bottleneck in the video engine, due to the way the frame pools are managed. The seeking performance is very good again, both for MJPEG and MP4. The modifications are in the following commits:

https://github.com/tlecomte/Joker/commit/9c455479f04c8a9bd28e7407e84196dc223f7fc2
https://github.com/tlecomte/Joker/commit/a3a2fd6a1d924bce2dac1d323863302c16c11cd8
https://github.com/tlecomte/Joker/commit/87e70f550ecd2352e9b4e962bd6f3ee6dcada2c6

I'll make a pull request for that in the following days. Thanks!
